### PR TITLE
Added dcterms:type validation for catalog items

### PR DESCRIPTION
**Pull request by SaiKireeti2905**
This pull request introduces infrastructure and code changes to support validation of dataset types using configurable allowed values. It adds a new `AllowedValuesValidator`, updates the application context and validator interfaces to support a YAML-based configuration, and provides comprehensive tests for the new validator. Additionally, it sets up a Neo4j database with the n10s plugin using Docker Compose, along with initialization scripts and configuration files.

**Infrastructure and Environment Setup:**

* Adds a `docker-compose.yml` file to define services for Neo4j (with n10s plugin), an initialization container, and the main application, including environment variables and health checks.
* Adds a `.env.example` file to document required environment variables for Neo4j credentials.
* Introduces a `neo4j-init.sh` script to initialize Neo4j with required constraints, n10s configuration, namespace prefixes, and an initial catalog node.

**Configuration and Validation Logic:**

* Adds a YAML configuration file (`allowed_values.yaml`) to define allowed values for dataset type validation.
* Updates the `Context` and validator interfaces to accept an `allowed_values_config_path` parameter, ensuring the path to the configuration is available throughout the application. [[1]](diffhunk://#diff-8d672cb396d7754e617f2866a76fd0c56f33e665e23ca31ca5bc69fb63071e97R14) [[2]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R41) [[3]](diffhunk://#diff-7ac9bf6ec2760b54f87fe8708d824890701016b8cdc48824f6b741f6669c84c7R163)
* Introduces a new `ConfigurationError` exception for configuration-related errors.

**Validation Implementation and Testing:**

* Implements the `AllowedValuesValidator` class, which loads and validates the allowed values configuration and enforces allowed values for dataset types.
* Adds comprehensive unit tests for `AllowedValuesValidator`, covering valid and invalid configurations, validation logic, and error handling. [[1]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1R4-R24) [[2]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1R144-R337)
* Updates test contexts and mocks to include the `allowed_values_config_path` parameter. [[1]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R197) [[2]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R272)This pull request introduces support for configurable allowed values validation for dataset entities, adds a robust YAML-based configuration system, and integrates Neo4j with n10s plugin initialization for local development. The main changes include adding a new `AllowedValuesValidator` with comprehensive error handling and tests, updating the application context and validator interfaces to support the new feature, and providing Docker-based infrastructure for Neo4j and service initialization.

**Allowed values validation system:**

* Added `AllowedValuesValidator` to `server/app/core/validators.py`, enabling YAML-configurable validation of RDF predicates for specific entity types, with detailed error handling and process-level caching.
* Expanded test coverage in `server/app/core/tests/test_validators.py` for the new validator, including tests for valid/invalid values, error cases, and configuration errors.
* Introduced a sample configuration file `server/app/core/allowed_values.yaml` for allowed values validation.
* Added the `ConfigurationError` exception for robust error reporting in configuration loading.

**Application and test integration:**

* Updated the validator service interface and implementation to accept `allowed_values_config_path`, and passed this through the context in `server/app/core/context.py`, `server/app/core/usecases.py`, and related tests. [[1]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R41) [[2]](diffhunk://#diff-7ac9bf6ec2760b54f87fe8708d824890701016b8cdc48824f6b741f6669c84c7R163) [[3]](diffhunk://#diff-8d672cb396d7754e617f2866a76fd0c56f33e665e23ca31ca5bc69fb63071e97R14) [[4]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R197) [[5]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R272)

**Infrastructure and local development:**

* Added `docker-compose.yml` to orchestrate Neo4j (with n10s plugin), a one-shot initialization service, and the main application, with environment variable support and health checks.
* Created `neo4j-init.sh` to initialize Neo4j with required constraints, graph config, and namespace prefixes, and to create an initial catalog node if missing.
* Added `.env.example` for local environment variable configuration.

**Additional improvements:**

* Improved imports and structure in `server/app/core/validators.py` and `server/app/core/tests/test_validators.py` for clarity and maintainability. [[1]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R5-R13) [[2]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1L6-R23)

These changes enable flexible, configuration-driven validation of datasets, simplify local setup with Docker, and ensure robust error handling throughout the validation workflow.This pull request introduces a new configuration-driven validation system for allowed RDF values, updates the validation pipeline to support it, and adds infrastructure for local development and testing with Neo4j and the n10s plugin. The most significant changes are the implementation of the `AllowedValuesValidator` (which loads rules from YAML), integration of this validator into the main validation flow, and comprehensive tests and configuration for allowed values. Additionally, Docker Compose and initialization scripts are provided for a consistent local environment.

**Allowed Values Validation System:**

* Introduced `AllowedValuesValidator` in `server/app/core/validators.py`, which loads allowed value rules from a YAML config, validates RDF nodes against these rules, and provides detailed error handling for misconfigurations. The validator caches rules for performance and is robust against malformed configs.
* Added `allowed_values_config_path` to the validation context and passed it through the relevant layers (`context.py`, `usecases.py`, and tests) to ensure the validator can access the config file. [[1]](diffhunk://#diff-8d672cb396d7754e617f2866a76fd0c56f33e665e23ca31ca5bc69fb63071e97R14) [[2]](diffhunk://#diff-7ac9bf6ec2760b54f87fe8708d824890701016b8cdc48824f6b741f6669c84c7R163) [[3]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R197) [[4]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R272) [[5]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R43)
* Added `ConfigurationError` exception for configuration-related errors in validation. [[1]](diffhunk://#diff-37c5d0e68273e1ddcee45e5a448abc8a090251cc874fe9f251d17c6036659eb2R37-R40) [[2]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R5-R13)

**Configuration and Testing:**

* Added a sample allowed values configuration at `server/app/core/allowed_values.yaml` and comprehensive unit tests for the validator, covering valid/invalid datasets, config errors, and edge cases. [[1]](diffhunk://#diff-250d7c4aa68269286adef25e1742ece4d29d630bd8ef3cf5cb6c062b7c8633b7R1-R7) [[2]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1L6-R23) [[3]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1R143-R325)

**Development Environment Setup:**

* Added a `docker-compose.yml` for local development, including Neo4j with the n10s plugin, a custom initialization container, and the main application service.
* Provided a `neo4j-init.sh` script to initialize Neo4j with required constraints, n10s setup, namespace prefixes, and a default catalog node.
* Added a `.env.example` file for environment variable configuration.

These changes collectively enable flexible, configuration-driven validation of RDF datasets, improve test coverage, and provide a reproducible local development environment.

**References:**  
[[1]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R74-R221) [[2]](diffhunk://#diff-8d672cb396d7754e617f2866a76fd0c56f33e665e23ca31ca5bc69fb63071e97R14) [[3]](diffhunk://#diff-7ac9bf6ec2760b54f87fe8708d824890701016b8cdc48824f6b741f6669c84c7R163) [[4]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1R143-R325) [[5]](diffhunk://#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R1-R66)This pull request introduces a new validator to ensure that catalog items have allowed types and updates the application and its tests to support this validation. It also adds a `docker-compose.yml` for local development, including Neo4j and initialization scripts. The most important changes are grouped below.

**Catalog Item Type Validation:**

* Added a new `CatalogItemTypeValidator` class in `server/app/core/validators.py` to check that all `dcat:Dataset` nodes have a `dcterms:type` property, and that the type is within an allowed list. Raises a `GraphValidationError` if not.
* Updated `BaseValidatorService` and its usage so that `allowed_catalog_item_types` is passed through context and used in validation. This affects the validator's constructor, the use case logic, and the API route for saving datasets. [[1]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R42) [[2]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0R157-R161) [[3]](diffhunk://#diff-7ac9bf6ec2760b54f87fe8708d824890701016b8cdc48824f6b741f6669c84c7R163) [[4]](diffhunk://#diff-fc765e76d19535ed30c10b8735dfc401878607f166d135458aa88c3672c041e9R166) [[5]](diffhunk://#diff-8d672cb396d7754e617f2866a76fd0c56f33e665e23ca31ca5bc69fb63071e97R14)
* Added comprehensive tests for `CatalogItemTypeValidator` in `server/app/core/tests/test_validators.py`, covering valid, invalid, and missing type scenarios.

**Configuration and Environment:**

* Added `allowed_catalog_item_types` to the `Settings` class in `server/app/settings.py` with default values for Dataset and Software types.
* Updated tests and test fixtures to include `allowed_catalog_item_types` in the context and expected calls. [[1]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R197-R200) [[2]](diffhunk://#diff-fa565e99dd60e743a94ada6223dede1a7bdd06492b419f4e7a227eab1653d264R275) [[3]](diffhunk://#diff-b1b4e6515aa2f8c7878305950b3cbd6f5f629e0056c76ce72a1f650b223ac846R91-R94)

**Local Development Setup:**

* Added a `docker-compose.yml` file that provisions Neo4j (with the n10s plugin), a custom initialization script, and the app server for local development.
* Added a `neo4j-init.sh` script to initialize Neo4j with required constraints, prefixes, and a default catalog node.

**Dependency and Import Updates:**

* Updated imports to include `DCTERMS` where needed for type validation. [[1]](diffhunk://#diff-e2d65a6914fd5ea702834e9b3186d2b8ad0a7311e56a7d755a5496c5d8c0f5b0L9-R9) [[2]](diffhunk://#diff-fbf97f3c0e51a3d1ad36af1cfd345829a1bc52090d9f3f457e88c4bc46fefdd1L10-R24)

### CODE REVIEW

File: server/app/core/validators.py [hunk 1]
```diff
@@ -69,6 +71,31 @@ def validate(self, entity: Graph) -> None:
...
+            if not type_values:
+                raise GraphValidationError(
+                    "catalog_item_type_error",
+                    "Catalog item must have a dcterms:type property.",
+                )
+            for type_val in type_values:
+                if type_val not in self.allowed_types:
+                    allowed_types_str = ", ".join(
+                        sorted(str(t) for t in self.allowed_types)
+                    )
```
Comment (RafyAmgadBenjamin) [inline server/app/core/validators.py]: This shouldn't be inside the for loops for performance optimization

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
+version: "3.8"
```
Comment (RafyAmgadBenjamin) [inline docker-compose.yml]: Have you tested this docker image locally 
are you sure that neosemantics plugin is working with this configuraiton? 

to be honest, I think this should be in different PR

File: server/app/core/validators.py [hunk 3]
```diff
@@ -69,6 +71,31 @@ def validate(self, entity: Graph) -> None:
             )
 
 
+class CatalogItemTypeValidator(IValidator):
+    def __init__(self, allowed_types: Sequence[str]) -> None:
```
Comment (RafyAmgadBenjamin) [inline server/app/core/validators.py]: This is not ontology agnostic, you define this class that mainly handle the type
per our discussion we said this implementation should be agnostic 

Here is my suggestion for implementation, for sure refine it as much needed 

```
import yaml
from rdflib import URIRef, Graph
from rdflib.namespace import RDF

class AllowedValuesValidator(IValidator):
    def __init__(self, config_path: str) -> None:
        with open(config_path) as f:
            config = yaml.safe_load(f)

        # Parse into: {scope: {predicate: {allowed_values}}}
        self.rules: dict[URIRef, dict[URIRef, set[URIRef]]] = {}
        for entry in config["validators"]:
            scope = URIRef(entry["scope"])
            self.rules[scope] = {
                URIRef(rule["predicate"]): {URIRef(v) for v in rule["allowed_values"]}
                for rule in entry["rules"]
            }

    def validate(self, entity: Graph) -> None:
        for scope, predicates in self.rules.items():
            nodes = list(entity.graph.subjects(RDF.type, scope))
            for node in nodes:
                for predicate, allowed in predicates.items():
                    values = list(entity.graph.objects(node, predicate))
                    if not values:
                        raise GraphValidationError(
                            "allowed_values_error",
                            f"Node of type {scope} must have predicate {predicate}.",
                        )
                    for val in values:
                        if val not in allowed:
                            raise GraphValidationError(
                                "allowed_values_error",
                                f"Invalid value {val} for {predicate}. "
                                f"Allowed: {', '.join(sorted(str(v) for v in allowed))}",
                            )

```


In this case the YAML file should be constructed something like this 

```
validators:
  - scope: "http://www.w3.org/ns/dcat#Dataset"
    rules:
      - predicate: "http://purl.org/dc/terms/type"
        allowed_values:
          - "http://example.org/types/TypeA"
          - "http://example.org/types/TypeB"
      - predicate: "http://purl.org/dc/terms/language"
        allowed_values:
          - "http://id.loc.gov/vocabulary/iso639-1/en"
          - "http://id.loc.gov/vocabulary/iso639-1/nl"

  - scope: "http://www.w3.org/ns/dcat#DataService"
    rules:
      - predicate: "http://purl.org/dc/terms/type"
        allowed_values:
          - "http://example.org/types/ServiceTypeA"
```

the path of th YAML file should be passed to this from env. variable and define also in the helm chart

File: server/app/core/validators.py [hunk 4]
```diff
@@ -131,9 +158,15 @@ def __init__(
...
         ontology_url: str | None = None,
+        allowed_catalog_item_types: Sequence[str] = (),
     ) -> None:
+        if not allowed_catalog_item_types:
+            raise RuntimeError(
+                "allowed_catalog_item_types is required and must not be empty."
+            )
         self.validators = [
             HasNodeValidator(rdf_type=DCAT.Dataset, single=True),
+            CatalogItemTypeValidator(allowed_catalog_item_types),
```
Comment (RafyAmgadBenjamin) [inline server/app/core/validators.py]: AS mentioned this one should point to the agnostic class 
AllowedValuesValidator

this is just an example for refrence, you can enhance this

**Review by RafyAmgadBenjamin — COMMENTED** [review]
You will also need to use a configmap and pass the YAML file, and update the related parts in the Helm chart

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+services:
+  neo4j:
+    image: neo4j:5.18.0
```
Comment (RafyAmgadBenjamin) [inline docker-compose.yml]: you can depend on our docker image for neo4j that already enable new semantics so you also use the same docker image used on deployment
we can make this as a seperate tast 
if so plz open this use story

File: neo4j-init.sh [hunk 6]
```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
```
Comment (RafyAmgadBenjamin) [inline neo4j-init.sh]: if you used our docker image for neo4j, you won't need this file

**Review by RafyAmgadBenjamin — APPROVED** [review]

File: docker-compose.yml (hunk 5 above)
Comment (SaiKireeti2905) [reply docker-compose.yml]: Removed docker-compose.yml from this PR — we'll set up the docker-compose with the ghcr.io/hiro-microdatacenters-bv/neo4j-semantics image as a separate user story so this PR stays focused on the validator.

File: neo4j-init.sh (hunk 6 above)
Comment (SaiKireeti2905) [reply neo4j-init.sh]: Agreed — removed neo4j-init.sh from this PR. Once we switch to the neo4j-semantics image in the follow-up task this script won't be needed.
